Architecture_Engine/Query/IsValidForDistanceCheck.cs

```diff
@@ -0,0 +1,100 @@
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Architecture.BuildersWork;
+using BH.oM.Architecture.Elements;
+using BH.oM.Geometry;
+using System.Collections.Generic;
+using System.Linq;
+using BH.oM.Spatial.ShapeProfiles;
+using BH.Engine.Geometry;
+using BH.oM.Geometry.CoordinateSystem;
+using System;
+
+namespace BH.Engine.Architecture
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        public static bool IsValidForDistanceCheck(this Opening opening1, Opening opening2, double maxDistance = double.NaN, double distanceTolerance = Tolerance.Distance, double angleTolerance = Tolerance.Angle)
+        {
+            // Check if the openings are parallel
+            if (opening1.CoordinateSystem.Z.IsParallel(opening2.CoordinateSystem.Z, angleTolerance) == 0)
+            {
+                return false;
+            }
+
+            // Check if the openigns are on the same plane/wall
+            Point centerPoint1 = opening1.CoordinateSystem.Origin;
+            Point centerPoint2 = opening2.CoordinateSystem.Origin;
+            Plane plane2 = new Plane { Origin = centerPoint2, Normal = opening2.CoordinateSystem.Z };
+
+            double distancePoint1Plane2 = centerPoint1.Distance(plane2);
+            double distanceDepths = opening1.Depth/2 + opening2.Depth/2;
+
+            if (distancePoint1Plane2 > distanceDepths + distanceTolerance)
+            {
+                return false;
+            }
+
+            // Check if the openings are closer than maxDistance
+            if (!double.IsNaN(maxDistance))
+            {
+                double diag1 = opening1.MaxDiagonal();
+                double diag2 = opening2.MaxDiagonal();
+
```

Comment:
I would add a check if neither `diag1` nor `diag2` is `double.NaN` (that is what is going to be returned in case of an opening with unusual shape being fed into `MaxDiagonal`). If the above is true, we need to calculate the distance because the pair cannot be excluded.

```
if (double.IsNaN(diag1) || double.IsNaN(diag2))
    return true;
```